AUI-3153 Add framework for not validating fields set in the instance …

gregory-bretall opened the pull request:
…as non-validatable

I made this fix at the recommendation of @jbalsas on https://issues.liferay.com/browse/PTR-424?focusedCommentId=1551840&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-1551840. The fix is reliant on the related pull for the Liferay code to fix the described issue. 

PR for that is here: https://github.com/jonmak08/liferay-portal/pull/1957

https://issues.liferay.com/browse/AUI-3153

jbalsas commented:
Hey @gregory-bretall, is there any particular reason why this is implemented in a different way from the proposal? 

We shouldn't limit this to a specific node. It would be better to provide a configurable selector, with a sensible default here, in AUI, that can simply be configured in a different way when used in Liferay Portal. That would make this more generic and flexible and save us from having to add support for multiple nodes or validation intereceptors in the future.

gregory-bretall commented:
I'm not sure I understand any real difference between your proposed implementation here (https://issues.liferay.com/browse/PTR-424?focusedCommentId=1551840&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-1551840) and how I implemented it. The main ways I changed it (adding a check if the relatedTarget is null to account for when you click and there is no target node) and using isEqualNode (because from what I could tell matches was not a valid function call) should give it essentially the same functionality. Could you please elaborate more on what you mean by your comment?

jbalsas commented:
Hey @gregory-bretall, yes there is.

You're testing against a specific node doing with `event.relatedTarget.getDOMNode().isEqualNode(nonValidatedField)` which you then need to set like

```
var nonValidatedField = document.getElementsByClassName('btn btn-cancel');

if (nonValidatedField) {
    formValidator.set('nonValidatedField', nonValidatedField[0]);
}
```

This presents lots of issues. 
- What if there is more than one `btn btn-cancel` on the page? Which one is the one that will allow you to cancel? Are you sure you always want to pick the `[0]` element?
- What if you want more than one way to get out of the form without validating? How do you turn your code to support multiple invalid-validation-triggers?
- What if the element is not yet in the page when you're querying for it? 

Here's the reference to the [matches](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches) method.

Please, take a look and follow our common approach:
- Add a `skipValidationTargetSelector` `ATTR` to `FormValidator` with a sensible default. For example: `'btn[type=cancel]'` 
- Use it to validate the `relatedTarget` if it exists
    - Additionally, we could add the condition that the node is contained within the form, but this could be too restrictive...
- In Portal, if necessary, set it to a different value, for example `btn[type=cancel],.btn.btn-cancel` at the beginning, when instantiating the `FormValidator`, not when binding the events.

/cc @antonio-ortega

gregory-bretall commented:
@jonmak08 @jbalsas I modified the code for this pull, please let me know what you think and if you need me to resend it on a new PR.

jbalsas commented on src/aui-form-validator/js/aui-form-validator.js:
```diff
@@ -1300,7 +1300,17 @@ var FormValidator = A.Component.create({
             var instance = this;
 
-            instance.validateField(event.target);
+            var skipValidationTargetSelectors = instance.get('skipValidationTargetSelectors');
```
This should be singular, `skipValidationTargetSelector` and it should be declared as a an `ATTR` of the `FormValidator` class.

gregory-bretall replied:
The reason I made it plural is because it can support multiple selectors. As for making it an ATTR, would you recommend making cancel the default in Alloy as well?

jbalsas replied:
It's actually a simple selector, it just happens that you can add `,` to it. Plural would mean this is an array of strings rather than a simple string and I'd expect to iterate through it rather than simply pass it to the [`matches`](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches) method that takes a `selectorString`.

jbalsas replied:
Regarding the default... it's a bit weird, but since `AlloyUI` actually relies on BS3 for a lot of things, it wouldn't be as crazy to set `button[type="cancel"],.btn.btn-cancel` as the default value. That would produce a bigger sweeping effect in Liferay Portal and reduce the need for changes over there.

I'd explore it. I'd be fine having it here.

jbalsas commented on src/aui-form-validator/js/aui-form-validator.js:
```diff
@@ -1300,7 +1300,17 @@ var FormValidator = A.Component.create({
+                skipValidationTargetSelectorFound = event.relatedTarget.getDOMNode().matches(skipValidationTargetSelectors);
+            }
+
+            if (!event.relatedTarget || !skipValidationTargetSelectorFound) {
```
This seems a bit redundant, isn't it? Shouldn't this do the same?

```javascript
// If the first clause is false, the second shouldn't be evaluated, and if it isn't then `event.relatedTarget` won't be undefined
if (!event.relatedTarget || event.relatedTarget.getDOMNode().matches(skipValidationTargetSelector)) {
    ... 
}
```

gregory-bretall replied:
Thanks for pointing that out, my approach above was a hold over from a more complex earlier solution I had dropped but I forgot to fully simplify the code

jbalsas replied:
👍

gregory-bretall commented:
@jbalsas I updated the PR again with your new suggestions. Let me know what you think.

jbalsas commented:
Looks good to me @gregory-bretall! Let's see what @jonmak08 thinks :)

jonmak08 commented:
Hey @gregory-bretall and @jbalsas,

Should the value of the ATTR `skipValidationTargetSelector` allow for an array of CSS selectors or should it just be one single string?

I only wonder because I could see it where you might want to have a few CSS selectors bypass the validation, but maybe I am wrong.

Let me know your thoughts.  Thanks!

jonmak08 commented:
Reviewed and Merged.  Thanks!
